sklearn/metrics/_regression.py

```diff
@@ -1188,24 +1239,33 @@ def r2_score(
         return float("nan")
 
     if sample_weight is not None:
-        sample_weight = column_or_1d(sample_weight)
-        weight = sample_weight[:, np.newaxis]
+        sample_weight = column_or_1d(sample_weight, dtype=dtype)
+        weight = sample_weight[:, None]
     else:
-        weight = 1.0
+        weight = xp.asarray([1.0], dtype=y_true.dtype, device=device_)
 
-    numerator = (weight * (y_true - y_pred) ** 2).sum(axis=0, dtype=np.float64)
-    denominator = (
-        weight * (y_true - np.average(y_true, axis=0, weights=sample_weight)) ** 2
-    ).sum(axis=0, dtype=np.float64)
+    numerator = xp.sum(weight * (y_true - y_pred) ** 2, axis=0, dtype=xp.float64)
+    denominator = xp.sum(
+        weight * (y_true - _average(y_true, axis=0, weights=sample_weight, xp=xp)) ** 2,
+        axis=0,
+        dtype=xp.float64,
+    )
 
-    return _assemble_r2_explained_variance(
+    result = _assemble_r2_explained_variance(
         numerator=numerator,
         denominator=denominator,
         n_outputs=y_true.shape[1],
         multioutput=multioutput,
         force_finite=force_finite,
+        xp=xp,
     )
 
+    result = input_xp.asarray(result, device=device_)
+    if result.size == 1:
+        return xp.reshape(result, (-1,))[0]
+
+    return result
```

Comment:
At a first look, this is black magic to me :)

First question, why do we need to call `input_xp.asarray`? I would expect the results to still be on the device, isn't it?

Then, the `reshape` with (-1,) is kind of obscure to me. Without digging, I would think this is to get the element of a 0d-array or something but I'll need to debug to understand. Maybe a comment would be helpful (or do I need to be more familiar with the array API?)

Reply:
> First question, why do we need to call `input_xp.asarray`? I would expect the results to still be on the device, isn't it?

Because towards the top of the function we might convert to numpy. Then perform some calculations with those arrays.

In general, it seems to be an emerging pattern that you end up calling `xp.asarray()` (or in this case `input_xp`) to make sure the result is in the same namespace as the input. I think another case would be where a numpy array or a numpy array api array is used as input. We pass both to Numpy functions and the nget Numpy arrays back. But we want to make sure we return a Numpy Array API array, if that was the input.



> Then, the `reshape` with (-1,) is kind of obscure to me.

Off the top of my head I also can't explain it. So maybe a comment is indeed useful

Reply:
For the last point, now that some time has passed I can't recall either. I bet some tests will fail with some array namespace if you remove this part. I might have spotted some differences between some libraries that output `(n,)`-shaped arrays, and others that output `(1, n)`-shaped arrays and must be squeezed afterward to get the expected dimension. And I prefer using `reshape` than squeeze, because squeezing `(1,)`-shaped arrays remove the dimension completely, while the output number of dimensions when reshaping is always consistent.

Reply:
I second @glemaitre 's point, also noted in another comment, this doesn't look right by just reading the code, and the only way to understand it would be to debug and come here and see  what's happening.

Reply:
For the first point, I also agree that converting back to the original namespace is weird. It means that we introduce a return dtype that is device specific, even if the intermediate computation is using numpy/CPU to be able to perform computation in float64 and the device does not support float64.

I think we should return `numpy` / `float64` values instead of attempting to convert back to the input namespace with a different dtype.